### PR TITLE
Add BeeWare Tutorial 1 support for GTK4

### PR DESCRIPTION
**Pull request by danyeaw**
This PR is a first step for #3069 through:

- Add a switch to turn on using GTK4 by setting the environmental variable `TOGA_GTK=4`
- Make minimal changes to allow running a Hello World app using GTK4
- Add a new CI test backend for linux-wayland-gtk4

Future PRs would implement libadwaita support and changes for additional widgets until we have a complete backend and then GTK3 can be removed.

Things left to do:

- [x] Determine why I am getting a libxapp error when trying to run the test suite with `TOGA_GTK=4` set, maybe XApp is GTK3 only for Cinnamon, MATE and Xfce
- [x] Update tests / mark them xfail if not migrated yet
- [x] Fix my co-authored statements to give @MuhammadMuradG credit
- [x] Add changelog entry

## PR Checklist:

- [x] All new features have been tested
- [x] All new features have been documented
- [X] I have read the **CONTRIBUTING.md** file
- [X] I will abide by the code of conduct

image

### CODE REVIEW

**Review by freakboy3742 — COMMENTED** [review]
I know this is still in draft, but I've flagged a couple of things that stood out to me in a quick look through the code.

Also - what are you considering the point of completion for this specific PR? The title suggests "hello world"... but I'm not sure how complex your intended hello world target is. Is it Tutorial 0? The BeeWare tutorial? Something else? To be clear - I'm totally on board with landing a PR that only works for a big subset of widgets so that we can avoid a monster PR that needs to live for months; I'm just trying to gauge how far you're thinking of going with this specific PR.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -317,6 +318,34 @@ jobs:
...
+
+            # Start Virtual X Server
+            echo "Start X server..."
+            Xvfb :99 -screen 0 2048x1536x24 &
+            sleep 1
+
+            # Start Window Manager
+            echo "Start window manager..."
+            # mutter is being run inside a virtual X server because mutter's headless
+            # mode is not compatible with Gtk
```
[the remark this answers is not in the document]
Comment (danyeaw) [reply .github/workflows/ci.yml:341]: Hi @freakboy3742, I can definitely investigate this either as part of this PR or as a separate one. Something like this should work with at least GTK4:

        env:
          XDG_RUNTIME_DIR: /tmp
        run: |
          eval $(dbus-launch --auto-syntax)
          DISPLAY=:99 MUTTER_DEBUG_DUMMY_MODE_SPECS=2048x1536 \
            mutter --wayland --no-x11 --sm-disable --headless toga &
          sleep 1

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -237,6 +237,7 @@ jobs:
         - "windows"
         - "linux-x11"
         - "linux-wayland"
```
[the remark this answers is not in the document]
Comment (danyeaw) [reply .github/workflows/ci.yml]: Thanks @freakboy3742, good point, I renamed the others to make them more explicit.

File: gtk/src/toga_gtk/libs/utils.py [hunk 3]
```diff
@@ -11,3 +11,7 @@ def gtk_text_align(alignment):
         CENTER: (0.5, Gtk.Justification.CENTER),
         JUSTIFY: (0.0, Gtk.Justification.FILL),
     }[alignment]
+
+
+def is_gtk3():
+    return Gtk.get_major_version() == 3
```
[the remark this answers is not in the document]
Comment (danyeaw) [reply gtk/src/toga_gtk/libs/utils.py]: I like this idea and I implemented the tuple in `gtk.py`.

File: .github/workflows/ci.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (danyeaw) [reply .github/workflows/ci.yml]: I tried to get this to work today. Although it does support running / testing a GTK app, our usage of Gdk.Screen / Gdk.Display doesn't seem to support a virtual display, and Toga throws a runtime error since the default_display returns None.

**Comment by danyeaw** [comment]
In the spirit of a minimal implementation, I am only running a few of the test modules (window, app, and desktop) with wayland-gtk4. We could then enable additional tests as we implement more modules. What do you think of this approach?

No rush on the feedback, next week would be completely fine. Cheers!

**Review by freakboy3742 — COMMENTED** [review]
>In the spirit of a minimal implementation, I am only running a few of the test modules (window, app, and desktop) with wayland-gtk4. We could then enable additional tests as we implement more modules. What do you think of this approach?

I can see how this approach would work, and I'm not *completely* opposed to it as an approach. However, if I had a choice, I'd prefer a "converging on 0%" approach, rather than a "converging on 100%" approach. That is, as part of this PR, we add a whole lot of explicit *exclusions*; when a subsequent PR adds support for a specific feature, that PR also deletes the testing exclusion.

That way the end state of the GTK porting effort is "all exclusions deleted", with the list of individual exclusions also working an effective TODO list. The approach you've used here would require the last PR to also delete the full list of "features that work" to revert to a "run all" configuration, which means it's a little harder to evaluate what is left to be done - and fairly easy for someone to add a new test file and forget to add the test to the GTK4 test list. It also means that code coverage testing is hidden until we reach the end-state, so it's difficult to evaluate if the PR enabling a specific feature has actually reached 100% coverage for the file implementing that feature.

The "converge on 0" approach is a little more invasive initially, but in a fairly straightforward way - adding a `if GTK_VERSION >= (4,0): pytest.skip()` in the constructor of each GTK probe implementation. It also means the initial test run is 95% skips, which is admittedly a little wasteful, but if the skip is early enough in the constructor process, then tests should run pretty quickly.

Speaking of coverage, the other thing that I noticed from the CI run is that we're going to add a new conditional-coverage qualifier - we already have [a bunch of these for Python versions and PIL availability]( https://github.com/beeware/toga/blob/main/pyproject.toml#L38); adding one based on the `TOGA_GTK` environment variable should be fairly straightforward.

**Comment by danyeaw** [comment]
I attempted to add the coverage exclusions, but it doesn't seem like the CI is picking it up. If you have a few minutes to see if you see something obvious that I missed that would be great, otherwise I'll try to reproduce it locally tomorrow.

**Comment by rmartin16** [comment]
Conditional coverage is a little more complex for Testbed. They need to be [injected](https://github.com/beeware/toga/blob/main/testbed/tests/testbed.py#L152) in to the app process.

**Comment by danyeaw** [comment]
I am getting a `Test suite didn't report a result` error with GTK4-Wayland midway through the tests consistently at `tests/widgets/test_mapview.py::test_flex_widget_size`. I am not getting this error locally. Any hints on troubleshooting this, maybe make use of the tmate debug option on the runner?

**Comment by freakboy3742** [comment]
> I am getting a `Test suite didn't report a result` error with GTK4-Wayland midway through the tests consistently at `tests/widgets/test_mapview.py::test_flex_widget_size`. I am not getting this error locally. Any hints on troubleshooting this, maybe make use of the tmate debug option on the runner?

I'm not 100% sure on the cause - but my guess is that it's trying to instantiate the web view that underpins the map view, failing, and leaving the test in a weird state. 

In terms of how to troubleshoot - `TOGA_GTK=4 briefcase dev --test -- tests/widgets/test_mapview.py` will launch the test suite, running *just* the mapview tests. Adding `--slow` to the end of of that command line also helps, as you then get to see all of stdout, plus any progress messages.

With that invocation, I've been able to reproduce the failure - it shows the app window, then shows a GTK "App is not responding" dialog, then crashes during `test_flex_widget_size` yielding a "Test suite didn't report a result". 

With some additional print statements in the widget, probe and test case to trace how far it was getting, I've worked out that the problem is that the call to `self.native.load_html()` in GTK MapView is failing. That means widget creation is failing, and everything falls in a heap from there.

The immediate observation that comes from this: putting the GTK-skip logic in the probe may not be sufficient. My first attempt at a brute-force fix was to just comment out the `load_html` call... but that doesn't work, because the test widget then relies on waiting for the widget initialisation to finish - which it never does, because the widget isn't fully initialised. So - even if we are skipping all the widget tests, the testbed suite must be able to *instantiate* all the widgets to be able to run at all.

That said - there's a limited number of special cases here. From my testing, `MapView`, `WebView`, and `OptionContainer` are the only widgets that can't be instantiated using *literally* the same code. I've pushed an update that should protect against these failures - with the added bonus that users who don't have WebKit2 installed can now run the test suite without a similar class of crash (they'll get an XFAIL for affected tests).

We *could* elevate that logic to *all* widgets... but I'm not sure that's worth it. Right now, many of the widgets will kinda-sorta-work, so I don't think we gain much by adding a bunch of code to explicitly kill widgets when they *might* work to first approximation. Protecting against the widgets with *known* failure modes should be enough; and even when we *do* fully support GTK4, the "protect against a missing WebKit2" handling will still be valuable for WebView and MapView.

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
We have an all green CI! Nice work!

I've flagged a couple of minor things inline (some of which are clarifications for my own benefit), but this looks pretty close to being ready to merge. Now we just need to draw the rest of the owl... 🤣 

Three final notes:
1. Let me know if any of the test-skipping changes I've made don't make sense, or if you think I've missed an edge case
2. I think I know the answer to this - but these changes look like there assuming GTK4 will *always* be running under Wayland. I'm assuming that's because X is almost as much of a dead end as GTK3 at this point? It wouldn't surprise me if this is true, but I'm not close enough to the Linux space to be 100% certain. 
3. It would be nice to revert the changes I made to OptionContainer before landing this PR. Those changes aren't *quite* the same case as WebView/MapView - those tests have a more complex "we actually need the widget to work" requirement to even make the testbed skip correctly; and there's a pre-existing edge case around not having WebKit installed that we are now able to catch. However, for OptionContainer, I was seeing a hard segfault on `test_focus_noop`. This is presumably because of an API that doesn't works differently on GTK4. I'd prefer to land this with that specific API call gated out (as we've done on other widgets), rather than the "isn't supported" check - but I also don't think it's worth spending a *lot* of time to diagnose the problem if it doesn't fall out quickly.

File: gtk/src/toga_gtk/dialogs.py [hunk 4]
```diff
@@ -220,16 +232,20 @@ def __init__(
...
+        save_icon = "_Save"
         super().__init__(
             title=title,
             filename=filename,
             initial_directory=initial_directory,
             file_types=file_types,
             multiple_select=False,
             action=Gtk.FileChooserAction.SAVE,
-            ok_icon=Gtk.STOCK_SAVE,
+            ok_icon=save_icon,
```
[the remark this answers is not in the document]
Comment (danyeaw) [reply gtk/src/toga_gtk/dialogs.py:243]: Correct, this was deprecated in GTK 3.10. You can see the recommendation here to either use the `document-save` icon or directly use the `_Save` label:
https://docs.gtk.org/gtk3/const.STOCK_SAVE.html

I think the label is closer to what the GTK4 dialogs will look like, so thought I would moving in this direction would be good.

**Comment by danyeaw** [comment]
Thanks so much for the changes, these look really great!

> 2. I think I know the answer to this - but these changes look like there assuming GTK4 will _always_ be running under Wayland.

Although there are still some X holdouts, Wayland is now default for most major distros. I was thinking that when we are ready to drop support for GTK3, that we drop support for X at the same time. This could of course be revisited when it is time to cross that road though.

> 3. It would be nice to revert the changes I made to OptionContainer before landing this PR.

I agree, I wrapped OptionContainer like the other widgets to allow the self.native creation, but stub out everything else. 👍

File: gtk/src/toga_gtk/screens.py [hunk 5]
```diff
@@ -21,8 +21,11 @@ def get_name(self):
         return self.native.get_model()
 
     def get_origin(self) -> Position:
-        geometry = self.native.get_geometry()
-        return Position(geometry.x, geometry.y)
+        if GTK_VERSION < (4, 0, 0):  # pragma: no-cover-if-gtk4
+            geometry = self.native.get_geometry()
+            return Position(geometry.x, geometry.y)
+        else:  # pragma: no-cover-if-gtk3
+            return Position(0, 0)
```
[the remark this answers is not in the document]
Comment (danyeaw) [reply gtk/src/toga_gtk/screens.py:28]: Yup, great catch!

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -330,14 +331,44 @@ jobs:
...
+          runs-on: "ubuntu-24.04"
+          env:
+            XDG_RUNTIME_DIR: "/tmp"
+          # The package list should be the same as in unix-prerequisites.rst, and the BeeWare
+          # tutorial, plus mutter to provide a window manager.
+          pre-command: |
+            sudo apt update -y
+            sudo apt install -y --no-install-recommends \
+              mutter pkg-config python3-dev libgirepository1.0-dev libcairo2-dev \
+              gir1.2-webkit-6.0 gir1.2-xapp-1.0 gir1.2-geoclue-2.0 gir1.2-flatpak-1.0 \
```
[the remark this answers is not in the document]
Comment (danyeaw) [reply .github/workflows/ci.yml:353]: Unfortunately, Webkit 6.0 is GTK4 and Webkit2 4.x is GTK3 only.

I added some brief docs now. I think we only want to advertise this for developers who want to help with GTK4 until things become much more mature.

File: gtk/src/toga_gtk/screens.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (danyeaw) [reply gtk/src/toga_gtk/screens.py:31]: Added a marker here. Yup, GTK4 leaves all screen movement / getting positions to the window manager.

File: gtk/src/toga_gtk/container.py [hunk 7]
```diff
@@ -8,15 +8,107 @@
...
 #######################################################################################
 
+if GTK_VERSION < (4, 0, 0):  # pragma: no-cover-if-gtk4
+    LayoutManager = object
+else:  # pragma: no-cover-if-gtk3
+    LayoutManager = Gtk.LayoutManager
 
-class TogaContainer(Gtk.Fixed):
+
+class TogaContainerLayoutManager(LayoutManager):  # pragma: no-cover-if-gtk3
```
[the remark this answers is not in the document]
Comment (danyeaw) [reply gtk/src/toga_gtk/container.py]: I added a conditional in front of the class now, so it isn't defined at all unless it is with GTK4.

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
I've pushed a couple of updates to documentation; but this is starting to look good, and it's enough to run BeeWare tutorial step 1 (not very exciting, but ITS ALIVE!!)

The only problem I've noticed in my testing is a regression on GTK3 - if you run a simple GTK3 example (Toga tutorial0 or tutorial1 will show it), you get a lot of GTK warnings:
```
(__main__.py:257361): Gtk-WARNING **: 10:13:07.685: Drawing a gadget with negative dimensions. Did you forget to allocate a size? (node box owner toga_gtk+container+TogaContainer)
```
(repeated dozens of times). I'm guessing this is related to the changes in Container, but it's not immediately obvious to me which part of the diff is causing this problem. It doesn't show up on BeeWare step 1, presumably because there's no content in the window other than the main box.

**Comment by danyeaw** [comment]
Hi @freakboy3742, thanks so much for the documentation updates - really great idea to encourage others to contribute to the GTK4 updates!

It looks like the issue was using a Gtk.Box instead of a Gtk.Frame for the TogaContainer. I decided the best course of action is to create separate TogaContainers for GTK3 and 4 so that they have separate inheritance. I think this will help make it easier to make changes to the TogaContainer as we progress with the update as well, without having to worry about unintentional issues with GTK3.

I also pushed one additional item to help cleanup the get window state logic which was getting a little hairy.

**Review by freakboy3742 — APPROVED** [review]
That looks like it's done the job - given there's a clearly different layout mechanisms in use between GTK 3 and 4, I think it makes sense to keep the two separate.

I've simplified the container implementation a little - instead of multiple "if GTK4" checks, I've condensed it all into one big "if GTK4; else" - but otherwise, I think this is good to go. Thanks for getting this big change rolling!

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -418,7 +449,8 @@ jobs:
       timeout-minutes: 15
       run: |
         ${{ matrix.briefcase-run-prefix }} \
-          briefcase run ${{ matrix.platform }} --log --test ${{ matrix.briefcase-run-args }} -- --ci
+          briefcase run ${{ matrix.platform }} --log --test ${{ matrix.briefcase-run-args }} \
+          -- ${{ matrix.briefcase-test-args }} --ci
```
Comment (freakboy3742) [inline .github/workflows/ci.yml:453]:
```suggestion
          briefcase run ${{ matrix.platform }} --log --test ${{ matrix.briefcase-run-args }} -- --ci
```

**Comment by danyeaw** [comment]
Thanks @freakboy3742!
